validator/src/validator.rs

```diff
@@ -382,13 +415,52 @@ where
                     equivocation_proofs,
                     prev_seed,
                     next_block_number,
-                    Self::PRODUCER_TIMEOUT,
+                    Duration::from_millis(Self::compute_micro_block_producer_timeout(
+                        lower_timestamp,
+                        head.timestamp(),
+                        amount_of_blocks_produced,
+                    )),
                     Self::BLOCK_SEPARATION_TIME,
                 ));
             }
         }
     }
 
+    /// Returns a timestamp and the number of blocks produced between this timestamp and the provided head block (both inclusive).
+    /// The number of skip blocks are filtered out in this time range as their timestamp is canonical thus not representative.
+    fn producer_timeout_block_range(
+        blockchain: &RwLockReadGuard<Blockchain>,
+        head: &Block,
+    ) -> (u64, u32) {
+        let blocks = blockchain.get_blocks(
+            &head.hash(),
+            Self::NUM_BLOCKS_FOR_TIMEOUT_CALCULATION,
+            false,
+            Direction::Backward,
+            None,
+        );
+
+        if let Ok(block_range) = blocks {
+            // Filter out skip blocks as they should not be part of the dynamic micro block producer timeout
+            let filtered: Vec<&Block> = block_range
+                .iter()
+                .filter(|block| !block.is_skip())
+                .collect();
+            if let Some(oldest_block) = filtered.last() {
+                return (oldest_block.timestamp(), filtered.len() as u32 + 1);
+            }
+        }
+
+        let genesis_block = blockchain
+            .get_block_at(Policy::genesis_block_number(), false, None)
+            .expect("genesis block must be present in the ChainStore");
+
+        (
+            genesis_block.timestamp(),
+            head.block_number() - genesis_block.block_number() + 1,
+        )
+    }
+
```

Comment:
This was the formula I was using before I started filtering skip blocks except I used the head timestamp instead of the current system time 😅

You have to keep the fallback on the genesis block. There are situations (e.g. a very short chain) where the `block_at(current_block_number - window_size)` can't return a block.